opentelemetry-sdk/src/opentelemetry/sdk/metrics/export/aggregate.py

```diff
@@ -0,0 +1,62 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+import threading
+
+
+class Aggregator(abc.ABC):
+    """Base class for aggregators.
+
+    Aggregators are responsible for holding aggregated values and taking a
+    snapshot of these values upon export (check_point).
+    """
+
+    def __init__(self):
+        self.current = None
+        self.check_point = None
+        self._lock = threading.Lock()
+
+    @abc.abstractmethod
+    def update(self, value):
+        """Updates the current with the new value."""
+
+    @abc.abstractmethod
+    def checkpoint(self):
+        """Stores a snapshot of the current value."""
+
+    @abc.abstractmethod
+    def merge(self, other):
+        """Combines two aggregator values."""
+
+
+class CounterAggregator(Aggregator):
+    """Aggregator for Counter metrics."""
+
+    def __init__(self):
+        super().__init__()
+        self.current = 0
+        self.check_point = 0
+
+    def update(self, value):
+        self.current += value
+
+    def checkpoint(self):
+        # TODO: Implement lock-free algorithm for concurrency
```

Comment:
to completely contradict myself earlier: is it worth implementing a lock-free algorithm here? checkpoint shouldn't be called particularly often, and lock-free algorithms are easy to get wrong and can often lead to worse performance.

Reply:
There is no specs for for how to implement `checkpoint` to be concurrent. There was talks that using a lock would be too slow, but I agree that lock-free algorithms are prone to error and might not guarantee performance. I think I might change this to a question, rather than a TODO.